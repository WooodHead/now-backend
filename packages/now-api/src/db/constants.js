/* eslint-disable import/prefer-default-export  */

export const TABLES = {
  ACTIVITY: 'now_template',
  DEVICE: 'now_device',
  EVENT: 'now_event',
  RSVP: 'now_rsvp',
  MESSAGE: 'now_messages',
  USER: 'now_user',
};
