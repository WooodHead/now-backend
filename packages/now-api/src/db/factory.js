/* eslint-disable import/no-extraneous-dependencies  */
import { Factory } from 'rosie';
import uuid from 'uuid/v4';
import { address, name, date, internet, random, lorem, hacker } from 'faker';
import randomEmoji from 'random-emoji';
import { LocalDate, ZoneId, LocalDateTime } from 'js-joda';
import {
  EVENT_INVITE_TYPE,
  APP_INVITE_TYPE,
} from '../schema/resolvers/Invitation';

const md5 = () =>
  random
    .hexaDecimal(32)
    .toLowerCase()
    .substring(2);

Factory.define('user')
  .attr('id', uuid)
  .attr('auth0Id', () => random.alphaNumeric(24))
  .attr('email', internet.email)
  .attr('firstName', name.firstName)
  .attr('lastName', name.lastName)
  .attr('bio', lorem.paragraph)
  .attr('location', () => `${address.city()}, ${address.stateAbbr()}`)
  .attr('preferences', {})
  .attr('birthday', () =>
    LocalDate.now()
      .minusYears(20)
      .toString()
  )
  .attr('photoId', md5)
  .attr(
    'photoPreview',
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAoCAYAAACM/rhtAAAAAXNSR0IArs4c6QAAAmRJREFUWAntmN1Kw0AQhU/8Fwv1F0FEKCgV9UbvvKkPIFJ8AB/C1/BZfAdF9EpBFEUEBQUp0l6opRVEi/Y4LIlx006aWHuRgTHpZHbmy9lssuh81g0dbF0dzPaNlgBGnaFEwUTBqAqY8e/vwM4OUCiYiO7Ytmfw+Bi4vRXXoUlW2wD396VhqRQGD2gL4N0dcH/fwYBGPSIWiwKq/atSsFbTlvudV60CJyduPOwU97hD7WcfH8D2NpBOA0tLwOIikM0CfX32fH/06AjgCjb29ga8vEg9E2t0bArYU8/gfod3vrcnztjcnMASeGoquMXBwe9rrMUb1lhTQBYZGwMeH91yVPXqSnx3FxgZEVgqPD8PDA5K7uWl3Jg7Us4IODvrj9p/twToL/X0BBweind3A5mMgJ6e+jPld5iFogIcH7c3skW5oG5uxG3XGQuzUFSrOAxgEJQ3Hjsgn8E4LcwU/4uCr68A348aUwHGrSDBtCqqAFMpYGBAc7/6HO1zqAJk27hVjB0waCVPTAC5HDA6qlePmdopVr0HWTBIwa0t+TYz5+EBODsTN9srxm2mVVANaFNwYcGFI8T0tPj6OvD8LKDn58D19c8NA3P/HNBxgM1NtrLb8DCwtibOHQy/y1T34gKoVIByWaB7e+3jTVStoH+KV1aAmRlTpvGxvx9gPp07I34KqWAzOFZVA3qnuKu+9vP5xlBBV6k8t2p0jalfM3wPDg1JydVVYHJSUz56jhqQragip2VjI3pjbQX1FLMgd8/Ly7JB1TaImuck/8CMKGGoZzBir5aGJ4AtyeYZlCjoEaOl0y+Le57bWRVobwAAAABJRU5ErkJggg=='
  );

Factory.define('device')
  .option('user', () => ({ id: uuid() }))
  .attr('token', uuid)
  .attr('type', () => random.arrayElement(['ios', 'android']))
  .attr('model', random.randomWord)
  .attr('userId', ['user'], ({ id }) => id);

Factory.define('rsvp')
  .option('user', () => ({ id: uuid() }))
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('action', 'add')
  .attr('userId', ['user'], ({ id }) => id)
  .attr('eventId', ['event'], ({ id }) => id);

Factory.define('appInvite')
  .option('inviter', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('type', APP_INVITE_TYPE)
  .attr('code', () => String(random.number({ max: 999999, min: 100000 })))
  .attr('expiresAt', () =>
    LocalDateTime.now()
      .plusDays(1)
      .toString()
  )
  .attr('inviterId', ['inviter'], ({ id }) => id);

Factory.define('eventInvite')
  .option('inviter', () => ({ id: uuid() }))
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('type', EVENT_INVITE_TYPE)
  .attr('code', () => String(random.number({ max: 999999, min: 100000 })))
  .attr('expiresAt', () =>
    LocalDateTime.now()
      .plusDays(1)
      .toString()
  )
  .attr('inviterId', ['inviter'], ({ id }) => id)
  .attr('eventId', ['event'], ({ id }) => id);

Factory.define('activity')
  .attr('id', uuid)
  .attr('activityDate', () =>
    date
      .soon(1)
      .toISOString()
      .substring(0, 10)
  )
  .attr('title', () => `${hacker.ingverb()} for ${hacker.noun()}`)
  .attr('emoji', () => randomEmoji.random({ count: 1 })[0].character)
  .attr('description', hacker.phrase);

Factory.define('todayActivity')
  .extend('activity')
  .attr('activityDate', () => new Date().toISOString().substring(0, 10));

Factory.define('event')
  .option('activity', () => ({
    id: uuid(),
    activityDate: date
      .soon(1)
      .toISOString()
      .substring(0, 10),
  }))
  .option('location', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('activityId', ['activity'], ({ id }) => id)
  .attr('locationId', ['location'], ({ id }) => id)
  .attr('limit', () => random.number({ max: 20, min: 5 }))
  .attr('time', ['activity'], ({ activityDate: activityDateString }) => {
    const activityDate = LocalDate.parse(activityDateString);
    return date
      .between(
        activityDate.atStartOfDay(),
        activityDate.atTime4(23, 59, 59, 999)
      )
      .toISOString();
  })
  .attr('timezone', () => random.arrayElement(ZoneId.getAvailableZoneIds()));

Factory.define('location')
  .attr('id', uuid)
  .attr('name', hacker.ingverb)
  .attr('address', address.streetAddress)
  .attr('city', address.city)
  .attr('state', address.stateAbbr)
  .attr('postalCode', address.zipCode)
  .attr('country', address.countryCode)
  .attr('neighborhood', hacker.noun)
  .attr('foursquareVenueId', md5)
  .attr(
    'location',
    () => `SRID=4326;POINT(${address.longitude()} ${address.latitude()})`
  );

Factory.define('message')
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('ts', () => date.recent(1).getTime())
  .attr('text', hacker.phrase)
  .attr('eventId', ['event'], ({ id }) => id)
  .attr('userId', ['user'], ({ id }) => id);

export default Factory;
