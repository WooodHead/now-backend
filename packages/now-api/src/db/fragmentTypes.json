{
  "__schema": {
    "types": [
      {
        "kind": "INTERFACE",
        "name": "Node",
        "possibleTypes": [
          { "name": "Activity" },
          { "name": "Event" },
          { "name": "Community" },
          { "name": "User" },
          { "name": "Rsvp" },
          { "name": "EventInvitation" },
          { "name": "Message" },
          { "name": "Location" },
          { "name": "AppInvitation" },
          { "name": "AdminMessage" }
        ]
      },
      {
        "kind": "INTERFACE",
        "name": "Connection",
        "possibleTypes": [
          { "name": "ActivityEventsConnection" },
          { "name": "CommunityUsersConnection" },
          { "name": "UserRsvpsConnection" },
          { "name": "CommunityEventsConnection" },
          { "name": "EventRsvpsConnection" },
          { "name": "EventMessagesConnection" },
          { "name": "RootActivitiesConnection" },
          { "name": "RootCommunitiesConnection" },
          { "name": "RootEventsConnection" },
          { "name": "RootRsvpsConnection" },
          { "name": "RootLocationsConnection" },
          { "name": "AppInvitationConnection" },
          { "name": "RootUsersConnection" },
          { "name": "AdminMessagesConnection" }
        ]
      },
      {
        "kind": "INTERFACE",
        "name": "Edge",
        "possibleTypes": [
          { "name": "ActivityEventsEdge" },
          { "name": "CommunityUsersEdge" },
          { "name": "UserRsvpsEdge" },
          { "name": "CommunityEventsEdge" },
          { "name": "EventRsvpsEdge" },
          { "name": "EventMessagesEdge" },
          { "name": "RootActivitiesEdge" },
          { "name": "RootCommunitiesEdge" },
          { "name": "RootEventsEdge" },
          { "name": "RootRsvpsEdge" },
          { "name": "RootLocationsEdge" },
          { "name": "AppInvitationEdge" },
          { "name": "RootUsersEdge" },
          { "name": "AdminMessagesEdge" }
        ]
      },
      {
        "kind": "INTERFACE",
        "name": "Invitation",
        "possibleTypes": [
          { "name": "EventInvitation" },
          { "name": "AppInvitation" }
        ]
      },
      {
        "kind": "UNION",
        "name": "FieldValue",
        "possibleTypes": [
          { "name": "StringFieldValue" },
          { "name": "IntFieldValue" },
          { "name": "DateTimeFieldValue" }
        ]
      }
    ]
  }
}
