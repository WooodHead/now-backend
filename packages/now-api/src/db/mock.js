/* eslint-disable import/no-extraneous-dependencies  */
import ApolloClient from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';

import schema from '../schema';
import loaders from './loaders';

export const USER_ID = 'aea68a98-591c-11e8-9cdb-872a15ebfd30';
export const mockPromise = (resolver, rejecter) =>
  rejecter ? Promise.reject(rejecter) : Promise.resolve(resolver);

export const mocks = {
  loadMember: jest.fn(),
};

const apolloCache = new InMemoryCache();

let isAdmin = false;

export const setAdmin = b => {
  isAdmin = b;
};

export const client = new ApolloClient({
  cache: apolloCache,
  link: new SchemaLink({
    schema,
    context: () => ({
      token: null,
      user: { id: USER_ID },
      loaders: loaders({ currentUserId: USER_ID }),
      scopes: isAdmin ? ['admin'] : [],
    }),
  }),
});

export const newUserClient = (currentUserAuth0Id, currentUserId) =>
  new ApolloClient({
    cache: apolloCache,
    link: new SchemaLink({
      schema,
      context: {
        token: null,
        currentUserAuth0Id,
        loaders: loaders({ currentUserId }),
      },
    }),
  });
