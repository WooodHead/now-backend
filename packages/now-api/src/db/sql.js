import knex from 'knex';
import { setPgTypes } from './types';
import knexConfig from '../../knexfile';

const config = knexConfig[process.env.NODE_ENV];
const sql = knex(config);

setPgTypes(sql);

export default sql;
export const connection = config.connection; // eslint-disable-line prefer-destructuring

export const genRandomUuid = () => sql.raw('gen_random_uuid()');
export const makepoint = (x, y) => sql.raw('st_makepoint(?, ?)', x, y);
export const now = () => sql.raw('now()');
