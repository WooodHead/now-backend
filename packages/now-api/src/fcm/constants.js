/* eslint-disable import/prefer-default-export */
export const NEW_EVENTS_TOPIC = 'new-events';
