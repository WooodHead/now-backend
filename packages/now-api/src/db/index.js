import AWS from 'aws-sdk';
import promisify from 'util.promisify';
import { rsvpId } from '../schema/util';
import { TABLES } from './constants';
import { isDev } from '../util';

const dynamoOpts = {};
if (isDev()) {
  dynamoOpts.logger = console;
}

const dynamoDb = new AWS.DynamoDB.DocumentClient(dynamoOpts);

const pGet = promisify(dynamoDb.get.bind(dynamoDb));
const pScan = promisify(dynamoDb.scan.bind(dynamoDb));
const pUpdate = promisify(dynamoDb.update.bind(dynamoDb));
const pPut = promisify(dynamoDb.put.bind(dynamoDb));
const pQuery = promisify(dynamoDb.query.bind(dynamoDb));

export const get = (table, key = {}) =>
  pGet({
    TableName: table,
    Key: key,
  }).then(response => response.Item);

export const put = (table, item, condition = undefined) =>
  pPut({
    TableName: table,
    Item: item,
    ConditionExpression: condition,
  });

export const update = (
  table,
  key,
  expr,
  values,
  names = undefined,
  condition = undefined
) =>
  pUpdate({
    TableName: table,
    Key: key,
    ExpressionAttributeValues: values,
    ExpressionAttributeNames: names,
    UpdateExpression: expr,
    ConditionExpression: condition,
    ReturnValues: 'ALL_NEW',
  });

export const updateDynamic = (table, key, values, condition = undefined) => {
  const expressions = [];
  const expressionNames = {};
  const expressionValues = {};
  Object.entries(values).forEach(([k, v], i) => {
    const attrName = `#f${i}`;
    const fieldName = `:f${i}`;
    expressions.push(`${attrName}=${fieldName}`);
    expressionNames[attrName] = k;
    expressionValues[fieldName] = v;
  });

  return update(
    table,
    key,
    `set ${expressions.join(', ')}`,
    expressionValues,
    expressionNames,
    condition
  );
};

export const scan = (table, filter = undefined) =>
  pScan({
    TableName: table,
    FilterExpression: filter,
  }).then(response => response.Items);

export const queryRaw = params => pQuery(params);
export const query = params => pQuery(params).then(response => response.Items);

export const getUserRsvpByEvent = (userId, eventId) =>
  get(TABLES.RSVP, { id: rsvpId(eventId, userId) });

export const getActivity = id => get(TABLES.ACTIVITY, { id });

export const getEvent = id =>
  pQuery({
    TableName: TABLES.EVENT,
    KeyConditionExpression: 'id = :id',
    ExpressionAttributeValues: { ':id': id },
    IndexName: 'id-index',
  }).then(response => response.Items[0]);

export { TABLES };
