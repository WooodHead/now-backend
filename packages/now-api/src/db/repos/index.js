// @flow
import type knex from 'knex';

import defaultSql from '../sql';
import { SQL_TABLES } from '../constants';
import { putInOrder } from '../../util';

type flat = string | number;

class Repo {
  constructor(table: string, sql: knex = defaultSql) {
    this.table = table;
    this.sql = sql;
  }

  get = (query: mixed = {}) =>
    this.sql(this.table)
      .where(query)
      .first();

  all = (query: mixed = {}) => this.sql(this.table).where(query);

  insert = (obj: mixed) => this.sql(this.table).insert(obj);

  update = ({ id, ...otherFields }: { id: flat }) =>
    this.sql(this.table)
      .where({ id })
      .update(otherFields);

  batch = (ids: Array<flat>) =>
    this.sql(this.table) // TODO: log batch savings
      .whereIn('id', ids)
      .then(batch => putInOrder(batch, ids));

  delete = (query: mixed) =>
    this.sql(this.table)
      .where(query)
      .delete();

  byId = (id: flat) => this.get({ id });

  withTransaction = (trx: ?knex): Repo =>
    trx ? new Repo(this.table, trx) : this;

  sql: knex;
  table: string;
}

export const User = new Repo(SQL_TABLES.USERS);
export const Device = new Repo(SQL_TABLES.DEVICES);
export const Activity = new Repo(SQL_TABLES.ACTIVITIES);
export const Event = new Repo(SQL_TABLES.EVENTS);
export const Location = new Repo(SQL_TABLES.LOCATIONS);
export const Rsvp = new Repo(SQL_TABLES.RSVPS);
export const RsvpLog = new Repo(SQL_TABLES.RSVP_LOG);
export const Message = new Repo(SQL_TABLES.MESSAGES);
export const Invitation = new Repo(SQL_TABLES.INVITATIONS);
export const EventUserMetadata = new Repo(SQL_TABLES.EVENT_USER_METADATA);
export const ServerMessages = new Repo(SQL_TABLES.SERVER_MESSAGES);
