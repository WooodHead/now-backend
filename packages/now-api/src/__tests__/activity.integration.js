import gql from 'graphql-tag';

import { client, setAdmin } from '../db/mock';
import { SQL_TABLES } from '../db/constants';
import sql from '../db/sql';
import factory from '../db/factory';
import { Activity } from '../db/repos';

const activities = factory.buildList('activity', 3);
const todayActivity = factory.build('todayActivity');
const event = factory.build('event', {}, { activity: todayActivity });

const truncateTables = () =>
  Promise.all([
    sql(SQL_TABLES.ACTIVITIES).truncate(),
    sql(SQL_TABLES.EVENTS).truncate(),
  ]);

beforeAll(() =>
  truncateTables().then(() =>
    Promise.all([
      sql(SQL_TABLES.ACTIVITIES).insert([...activities, todayActivity]),
      sql(SQL_TABLES.EVENTS).insert(event),
    ])
  ));
afterAll(() => {
  setAdmin(false);
  return truncateTables();
});

const ALL_ACTIVITIES_QUERY = gql`
  {
    allActivities {
      edges {
        node {
          id
          title
          description
          activityDate
          emoji
        }
      }
    }
  }
`;

describe('activity', () => {
  it("won't do allActivities by default", async () => {
    const results = client.query({ query: ALL_ACTIVITIES_QUERY });
    expect.assertions(1);
    return expect(results).rejects.toEqual(
      new Error('GraphQL error: You must be an admin.')
    );
  });

  it('return allActivities when admin', async () => {
    setAdmin(true);
    const results = client.query({
      query: ALL_ACTIVITIES_QUERY,
    });
    const { data } = await results;
    expect(data).toHaveProperty('allActivities');

    expect(data.allActivities.edges).toHaveLength(4);
    expect(data.allActivities.edges).toEqual(
      expect.arrayContaining(
        [...activities, todayActivity].map(a =>
          expect.objectContaining({
            __typename: 'RootActivitiesEdge',
            [Symbol('id')]: expect.anything(),
            node: expect.objectContaining({
              __typename: 'Activity',
              [Symbol('id')]: `Activity:${a.id}`,
              id: a.id,
              activityDate: a.activityDate,
              description: a.description,
              emoji: a.emoji,
              title: a.title,
            }),
          })
        )
      )
    );
  });

  it('gets single activity', async () => {
    const res = client.query({
      query: gql`
        query getActivity($id: ID!) {
          activity(id: $id) {
            id
            title
            emoji
            description
            activityDate
          }
        }
      `,
      variables: { id: activities[0].id },
    });

    const { data } = await res;
    expect(data).toMatchObject({
      activity: {
        __typename: 'Activity',
        [Symbol('id')]: `Activity:${activities[0].id}`,
        ...activities[0],
      },
    });
  });

  it("gets today's activity", async () => {
    const res = client.query({
      query: gql`
        {
          serverMessages {
            noActivityTitle
            noActivityMessage
          }
          todayActivity {
            id
            title
            emoji
            description
            activityDate
            events {
              edges {
                node {
                  id
                }
              }
            }
          }
        }
      `,
    });

    const { data } = await res;
    expect(data).toMatchObject({
      serverMessages: {
        noActivityTitle: 'Sorry, no Meetup today!',
        noActivityMessage:
          "We're either on vacay or planning something extra special for you.",
      },
      todayActivity: {
        __typename: 'Activity',
        [Symbol('id')]: `Activity:${todayActivity.id}`,
        ...todayActivity,
        events: {
          edges: expect.arrayContaining([
            expect.objectContaining({
              __typename: 'ActivityEventsEdge',
              [Symbol('id')]: `$Activity:${todayActivity.id}.events.edges.0`,
              node: expect.objectContaining({
                __typename: 'Event',
                id: event.id,
                [Symbol('id')]: `Event:${event.id}`,
              }),
            }),
          ]),
        },
      },
    });
  });

  it('creates an activity', async () => {
    const activity = factory.build('activity');
    const { title, description, activityDate, emoji } = activity;

    const res = await client.mutate({
      mutation: gql`
        mutation create($input: CreateActivityInput) {
          createActivity(input: $input) {
            activity {
              id
            }
          }
        }
      `,
      variables: {
        input: {
          title,
          description,
          activityDate,
          emoji,
        },
      },
    });

    const {
      data: {
        createActivity: {
          activity: { id },
        },
      },
    } = res;

    const dbActivity = await Activity.byId(id);
    dbActivity.activityDate = dbActivity.activityDate.toString();

    expect(dbActivity).toMatchObject({
      id,
      title,
      description,
      activityDate: expect.stringContaining(activityDate),
      emoji,
    });
  });
});
