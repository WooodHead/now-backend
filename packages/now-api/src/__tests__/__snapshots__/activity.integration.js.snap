// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activity creates an activity 1`] = `Object {}`;

exports[`activity creates an activity 2`] = `"data:image/jpeg;base64,/9j/2wBDAAoHBwgHBgoICAgLCgoLDhgQDg0NDh0VFhEYIx8lJCIfIiEmKzcvJik0KSEiMEExNDk7Pj4+JS5ESUM8SDc9Pjv/2wBDAQoLCw4NDhwQEBw7KCIoOzs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozv/wAARCAAoADwDASIAAhEBAxEB/8QAGwAAAgMBAQEAAAAAAAAAAAAABAUAAgMGAQf/xAAxEAACAQMCBAQFAgcAAAAAAAABAgMABBEFIRITMUEGImGBFEJRcbEyoRVikaLB0fD/xAAZAQEBAQADAAAAAAAAAAAAAAADBAIAAQX/xAAiEQACAgEEAgMBAAAAAAAAAAABAgARAwQSEyExURRBYcH/2gAMAwEAAhEDEQA/AC5fET3Mhh0y1edxtkLxY+/Ye5qieHdY1HmXk7RRyAYwCC3oOI+VfYGukt7QsBDawgKvyooVV/wKXXuu6fp2oT6XqFtJMsTqJOHDKW64x9BkffFcydCybm8YLGlEHsdIt7GVnaJzcKApklHmyOuM9KOLH6UuXxNp93fyxFpY24zlpFwOvf6UwKl8hMbDLMRsBWt6Ktzrhys+2u4fbS6fFBIW4pJkHGSdhgdeH7Uolu5ZZll5aJHK2FQZyP8AdLLz4m+uytpO0VumS7j5wNj7dv604srX+IWLIGYXEfy46EffrmonyNe5Z6WLTJRV+/5K80qxVgQR1H0qwnGKwbzpiVuW8QAzj9Xp7UOHLDIPpVGHMMg/ZFqdMcJseJpe+NNR0iSe0022ilmaQKHlQvjbBOARsN65XW7mW9uCbm6h+Ou34n4fKo26itpmllkuHuLiWOVG87Rw5zkdcdtqEtba2WcPbRuZHGHurw4UZ6DcfijLXLMeNVqpSe3jtbbnTSKjlshhJnbGMH/vzXfw24tvD8CQlkThC8Wc7HYn87186n06xsJGu7qeK5mkyVCZKtv1A6Y2I2p74U1nWp+G2VYZYWDlBOvRR2z75onWxHV6Md3Gl/FX4W1CCF2QFo2xwKvVQOm+1EajcvZTKbRg1wqniQDPF/KMd6pLqmpuY7UJDbhpBGzjB4cgnIPqRjf61no2nS2uq3V1Ncc0KNpCSBjG4H70VEdxQbsCCNrsOoQhX06456khozCQQe5z3oq1teZAJAWAffHcff1rTTAI3u7iXIgQnztsCQSCB+1GaWoFnxFSA7FlB7DtTYAN8k1l8fn7nL2/Il8UNbyxkxTQcKb43XfO3ToadTeHLG7jCyCcjIO8zMNumxyKlSmxgFZJnZg/RlLvwzFcoAHjLKMBpIhn+3Fe2fh+O1n+Ie5d5lTgjKjhWMeg7+9SpW9oh82T3N5bXnMVuLUTAjHEsvCCOu4NaNZajcW/JieG0iB/QAWJHqalSi4li/Kye5qmjSSMvxlxzUXcQonBGPami221SpWlAHiC7s5tjP/Z"`;
